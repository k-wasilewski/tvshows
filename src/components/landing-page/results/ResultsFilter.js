import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import FormControl from '@material-ui/core/FormControl';
import Select from '@material-ui/core/Select';
import {styles} from "../../../styles/styles";

const useStyles = makeStyles((theme) => ({
    formControl: {
        margin: theme.spacing(1),
        minWidth: 120,
    },
    selectEmpty: {
        marginTop: theme.spacing(2),
    },
}));

export default function SimpleSelect(props) {
    const materialUiClasses = useStyles();
    const classes = styles();
    const [day, setDay] = React.useState('');

    const handleChange = (event) => {
        setDay(event.target.value);
        props.setDay(event.target.value);
    };

    return (
        <div className={classes.dashboard}>
            <span className={classes.filterLabel}>Filtruj:</span>
            <FormControl className={`${materialUiClasses.formControl} ${classes.formControl}`}>
                <InputLabel id='demo-simple-select-label'>Dzień tygodnia</InputLabel>
                <Select
                    labelId='demo-simple-select-label'
                    id='demo-simple-select'
                    value={day}
                    onChange={handleChange}
                >
                    <MenuItem value={1}>Poniedziałek</MenuItem>
                    <MenuItem value={2}>Wtorek</MenuItem>
                    <MenuItem value={3}>Środa</MenuItem>
                    <MenuItem value={4}>Czwartek</MenuItem>
                    <MenuItem value={5}>Piątek</MenuItem>
                    <MenuItem value={6}>Sobota</MenuItem>
                    <MenuItem value={7}>Niedziela</MenuItem>
                </Select>
            </FormControl>
            <button onClick={props.doReset}>Resetuj filtr</button>
        </div>
    );
}