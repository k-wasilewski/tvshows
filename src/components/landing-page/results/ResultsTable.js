import React from 'react';
import { withStyles } from '@material-ui/core/styles';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableContainer from '@material-ui/core/TableContainer';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import Paper from '@material-ui/core/Paper';
import Tooltip from '@material-ui/core/Tooltip';
import { useHistory } from 'react-router-dom';
import {setDetailedResult} from "../../../redux/actions";
import {connect} from "react-redux";
import {
    TABLE_HEAD_STYLE,
    TABLE_ROW_STYLE,
    TABLE_STYLE,
    TABLE_CONTAINER_STYLE,
    TABLE_CELL_STYLE
} from '../../../styles/styles';

const mappedGenres = (genres) => {
    if (genres.length===1) return genres;
    else {
        return genres.map(genre => {
            if (genres.indexOf(genre)!==genres.length-1)
                return genre+=', ';
            else return genre;
        });
    }
};

const StyledTableHeadRow = withStyles((theme) => (TABLE_HEAD_STYLE))(TableRow);

const StyledTableRow = withStyles((theme) => (TABLE_ROW_STYLE))(TableRow);

const StyledTable = withStyles((theme) => (TABLE_STYLE))(Table);

const StyledTableContainer = withStyles((theme) => (TABLE_CONTAINER_STYLE))(TableContainer);

const StyledTableCell = withStyles((theme) => (TABLE_CELL_STYLE))(TableCell);

export function ResultsTable(props) {
    const [sortedResults, setSortedResults] = React.useState([]);
    const history = useHistory();

    React.useEffect( () => {
        setSortedResults(props.results.sort((a, b) => b.score - a.score));
    }, [props.results]);

    function rowClicked(event, result) {
        props.setDetailedResult(result);
        history.push('/details');
    }

    if (sortedResults.length===0) return <React.Fragment />
    else return (
        <StyledTableContainer component={Paper}>
            <StyledTable aria-label='results table'>
                <TableHead>
                    <StyledTableHeadRow>
                        <StyledTableCell>Ocena</StyledTableCell>
                        <StyledTableCell align='right'>Tytuł</StyledTableCell>
                        <StyledTableCell align='right'>Gatunki</StyledTableCell>
                        <StyledTableCell align='right'>Data premiery</StyledTableCell>
                    </StyledTableHeadRow>
                </TableHead>
                <TableBody>
                    {sortedResults.map((result) => (
                        <Tooltip key={`tooltip-${result.score}${result.show.name}`}
                                 title={`Szczegóły serialu ${result.show.name}`} placement='top-end'>
                            <StyledTableRow key={result.score+result.show.name}
                                            onClick={event => rowClicked(event, result)}>
                                <StyledTableCell component='th' scope='row'>
                                    {result.score}
                                </StyledTableCell>
                                <StyledTableCell align='right'>
                                    {result.show.name}
                                </StyledTableCell>
                                <StyledTableCell align='right'>
                                    {mappedGenres(result.show.genres)}
                                </StyledTableCell>
                                <StyledTableCell align='right'>
                                    {result.show.premiered}
                                </StyledTableCell>
                            </StyledTableRow>
                        </Tooltip>
                    ))}
                </TableBody>
            </StyledTable>
        </StyledTableContainer>
    );
};

const mapDispatchToProps = {
    setDetailedResult
};

export default connect(null, mapDispatchToProps)(ResultsTable);