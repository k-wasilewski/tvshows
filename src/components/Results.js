import React, { Component } from 'react';
import '../css/App.css';
import {setResults} from "../redux/actions";
import {connect} from "react-redux";
import ResultsTable from "./ResultsTable";
import ResultsFilter from "./ResultsFilter";
import mapDayNoToName from "../functions/mapDayNoToName";

class Results extends Component {
    constructor(props) {
        super(props);

        this.state = {
            filteredResults: []
        };
    }

    filterByDay = (day) => {
        const dayName = mapDayNoToName(day);

        const filteredResults = this.props.results.filter(result =>
            result.show.schedule.days.includes(dayName)
        );

        this.setState({filteredResults: filteredResults});
    }

    render() {
        return (
            <div className="App">
                <ResultsFilter setDay={this.filterByDay} />
                <ResultsTable results={(this.state.filteredResults.length===0) ?
                    this.props.results
                    :
                    this.state.filteredResults} />
            </div>
        );
    };
};

function mapStateToProps(state) {
    return {
        results: state.setResultsReducer.results,
    };
};

const mapDispatchToProps = {
    setResults
};

export default connect(mapStateToProps, mapDispatchToProps)(Results);
