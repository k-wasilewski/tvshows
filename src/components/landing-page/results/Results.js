import React from 'react';
import {connect} from 'react-redux';
import ResultsTable from './ResultsTable';
import ResultsFilter from './ResultsFilter';
import mapDayNoToName from '../../../functions/mapDayNoToName';
import 'babel-polyfill';

export class Results extends React.PureComponent {
    constructor(props) {
        super(props);
        this.loading=false;

        this.state = {
            filteredResults: []
        };

        this.filterByDay = this.filterByDay.bind(this);
        this.resetFilter = this.resetFilter.bind(this);
    }

    getSnapshotBeforeUpdate(prevProps, prevState) {
        this.loading=true;
        return null;
    }

    componentDidUpdate(prevProps, prevState, snapshot) {
        if (this.state.filteredResults===prevState.filteredResults) {
            this.setState({filteredResults: []});
            this.loading=false;
        }
    }

    filterByDay(day) {
        const dayName = mapDayNoToName(day);

        const filteredResults = this.props.results.filter(result =>
            result.show.schedule.days.includes(dayName)
        );

        this.setState({filteredResults: filteredResults});
    }

    resetFilter() {
        this.setState({filteredResults: this.props.results});
    }

    render() {
        const results = (this.state.filteredResults.length===0) ?
            this.props.results
            :
            this.state.filteredResults;

        const resultsFilter = (results.length===0) ?
            null
            :
            <ResultsFilter setDay={this.filterByDay} doReset={this.resetFilter} />;

        return (
            <React.Fragment>
                {resultsFilter}
                {(this.loading) ? 'loading' : null}
                <ResultsTable results={results} />
            </React.Fragment>
        );
    };
};

function mapStateToProps(state) {
    return {
        results: state.setResultsReducer.results,
    };
};

export default connect(mapStateToProps, null)(Results);
