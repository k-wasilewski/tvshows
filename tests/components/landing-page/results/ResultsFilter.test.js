import React from 'react';
import renderer from 'react-test-renderer';
import {Results} from "../../../../src/components/landing-page/results/Results";
import ResultsFilter from "../../../../src/components/landing-page/results/ResultsFilter";
import {configure, mount} from 'enzyme';
import Adapter from "enzyme-adapter-react-16";
import store from "../../../../src/redux/store";
import {Provider} from "react-redux";

describe("ResultsFilter rendering specification", () => {
    it('ResultsFilter is rendered', () => {
        const component = renderer.create(
            <ResultsFilter/>
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});

describe("ResultsFilter functional specification", () => {
    it('is only rendered when redux props.results are not empty', (done) => {
        configure({adapter: new Adapter()});

        const testFilteredResultsValue = {
            show: {
                name: 'sample name',
                genres: ['sample genre']
            },
            score: 2
        };

        const component = mount(
            <Provider store={store}>
                <Results results={[testFilteredResultsValue]}/>
            </Provider>
        );

        expect(component.find(ResultsFilter).exists()).toBeTruthy();
        component.setProps({children: <Provider store={store}><Results results={[]}/></Provider>});
        component.update();
        setTimeout(function () {
            expect(component.find(ResultsFilter).exists()).toBeFalsy();
            component.unmount();
            done();
        }, 500);
    });
});