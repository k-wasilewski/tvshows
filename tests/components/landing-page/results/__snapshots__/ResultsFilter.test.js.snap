// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultsFilter rendering specification ResultsFilter is rendered 1`] = `
<div
  className="makeStyles-dashboard-12"
>
  <h2
    className="MuiTypography-root makeStyles-filterLabel-13 MuiTypography-body1"
  >
    Filtruj:
  </h2>
  <div
    className="MuiFormControl-root makeStyles-formControl-1 makeStyles-formControl-14"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
      data-shrink={false}
      id="demo-simple-select-label"
    >
      Dzie≈Ñ tygodnia
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
    >
      <div
        aria-haspopup="listbox"
        aria-labelledby="demo-simple-select-label demo-simple-select"
        className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
        id="demo-simple-select"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="button"
        tabIndex={0}
      >
        <span
          dangerouslySetInnerHTML={
            Object {
              "__html": "&#8203;",
            }
          }
        />
      </div>
      <input
        aria-hidden={true}
        className="MuiSelect-nativeInput"
        onAnimationStart={[Function]}
        onChange={[Function]}
        required={false}
        tabIndex={-1}
        value=""
      />
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSelect-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      Resetuj filtr
    </span>
  </button>
</div>
`;
